### PR TITLE
Ensure same lock is used to read and write Dictionary in AppContext

### LINKED ISSUE #2727: Fix AppContext locking on 2 different objects
A similar fix to the one in [CoreFx](https://github.com/dotnet/corefx/pull/5517) needs to be made here.

@JonHanna let me know if you want to make the same change here as well, otherwise I'll go ahead and do it.

### PR DESCRIPTION
**Pull request by JonHanna**
Fixes #2727

### CODE REVIEW

File: src/mscorlib/src/System/AppContext/AppContext.cs [hunk 1]
```diff
@@ -65,79 +64,83 @@ public static bool TryGetSwitch(string switchName, out bool isEnabled)
...
+                //    Example: This will prevent us from trying to find overrides for values that we don't have in the dictionary
+
+                // 1. The value has an override specified.
+                bool overrideValue;
+                if (AppContextDefaultValues.TryGetSwitchOverride(switchName, out overrideValue))
+                {
+                    isEnabled = overrideValue;
+
+                    // Update the switch in the dictionary to mark it as 'checked for override'
+                    SetSwitch(switchName, isEnabled);
```
Comment (AlexGhiondea) [inline src/mscorlib/src/System/AppContext/AppContext.cs]: I would just explicitly set the value here. Calling SetSwitch is going to go through a lot more checks and code before it sets the value.

File: src/mscorlib/src/System/AppContext/AppContext.cs [hunk 2]
```diff
@@ -65,79 +64,83 @@ public static bool TryGetSwitch(string switchName, out bool isEnabled)
...
+                {
                     return true;
                 }
-                else
+
+                // 3. The switch has a valid value, but we need to check for overrides.
+                // Regardless of whether or not the switch has an override, we need to update the value to reflect
+                // the fact that we checked for overrides. 
+                bool overrideValue;
+                if (AppContextDefaultValues.TryGetSwitchOverride(switchName, out overrideValue))
```
Comment (AlexGhiondea) [inline src/mscorlib/src/System/AppContext/AppContext.cs]: Without the lock, this code could end up executing more than once. I think this should stay inside the lock as the call it makes is expensive and we should try and only do it once.

File: src/mscorlib/src/System/AppContext/AppContext.cs [hunk 3]
```diff
@@ -65,79 +64,83 @@ public static bool TryGetSwitch(string switchName, out bool isEnabled)
...
-                                                    | SwitchValueState.HasLookedForOverride;
-
-                        return true;
-                    }
+                    // we found an override!
+                    isEnabled = overrideValue;
+                }
+
+                // Update the switch in the dictionary to mark it as 'checked for override'
+                SetSwitch(switchName, isEnabled);
```
Comment (AlexGhiondea) [inline src/mscorlib/src/System/AppContext/AppContext.cs]: I would just explicitly set the value here. Calling SetSwitch is going to go through a lot more checks and code before it sets the value.

File: src/mscorlib/src/System/AppContext/AppContext.cs (hunk 2 above)
Comment (JonHanna) [inline src/mscorlib/src/System/AppContext/AppContext.cs]: Yes. That occurred to me the moment I turned off my computer. Will replace this in about ten minutes.

**Comment by AlexGhiondea** [comment]
@JonHanna thanks for doing this!

Unfortunatelly, what you see here is not the entire code that uses AppContext and makes some assumptions about they way it is implemented.

With that in mind, for this change, I would like to keep it as surgical as possible in order to not break some of the assumptions made by the rest of the code.

 I know it is difficult to fix something when you can't see all the parts.

File: src/mscorlib/src/System/AppContext/AppContext.cs (hunk 2 above)
Comment (AlexGhiondea) [inline src/mscorlib/src/System/AppContext/AppContext.cs]: Thank you! :) No rush!

**Comment by JonHanna** [comment]
I think the code in that file should have been enough to realise that that was a bad commit. I broke my rule about always waiting a few minutes before clicking "Create Pull Request", and I never don't regret it when I break that rule.

A more restrained update has been made.

**Comment by AlexGhiondea** [comment]
Thank you for the update! LGTM!

I am going to work on getting some of our tests for AppContext into the open to try and help with validating the changes to it!

File: src/mscorlib/src/System/AppContext/AppContext.cs [hunk 4]
```diff
@@ -157,11 +156,13 @@ public static void SetSwitch(string switchName, bool isEnabled)
             if (switchName.Length == 0)
                 throw new ArgumentException(Environment.GetResourceString("Argument_EmptyName"), "switchName");
 
-            lock (s_syncLock)
+            SwitchValueState switchValue = (isEnabled ? SwitchValueState.HasTrueValue : SwitchValueState.HasFalseValue)
+                                            | SwitchValueState.HasLookedForOverride
```
Comment (AlexGhiondea) [inline src/mscorlib/src/System/AppContext/AppContext.cs]: there is a missing trailing ';' which is causing the build break.

**Comment by JonHanna** [comment]
Test Ubuntu x64 Release Build and Test

**Pull request merged**